### DISCUSSION TITLE
Multi-tenant

### DISCUSSION THREAD

**Discussion by arsewizz** [supabase/supabase#1615] [Questions]
is there a way to create a database with the possibility of Multi-tenant ?

**Comment by steve-chavez** [comment]
This depends on how you'd like to structure your database. Basically, there are three ways:

1. Separate dbs for each tenant
2. Same db but separate schemas for each tenant
3. Same db and same schema, tenants separated by RLS policies

(Check this [article](https://web.archive.org/web/20080510114523/https://msdn.microsoft.com/en-us/library/aa479086.aspx) for knowing the types of multitenancy.)

I think 3 would be the most straightforward way for now(2 may be possible later). You could have a `tenant_id` and base your RLS policies on that, check this [AWS blog post](https://aws.amazon.com/blogs/database/multi-tenant-data-isolation-with-postgresql-row-level-security/) for more details on how to do it.

Option 1 seems bad for scalability. Since a connection is tied to a database, if you have many tenants you'll need to maintain a lot of connections(and in postgres this is bad because 1 connection = 1 process).

---

There was a previous discussion on: https://github.com/supabase/supabase/discussions/957#discussioncomment-548934

**Reply by arsewizz** [reply]
So do you think it's possible to do it with supabase ?

**Reply by pichichi91** [reply]
I think this should be perfectly possible - I'd guess you'd need to just link users to a certain tenant and then add that tenant to row-level security, as mentioned above.

Could be a nice idea to have an SaaS / multitenancy example available for supabase - on the other hand the to do list example already has some multi tenancy functionality but for simple users instead a whole group of users combinend to a tenant:

https://github.com/supabase/supabase/tree/master/examples/nextjs-todo-list

**Reply by KhaledGabr** [reply]
@pichichi91 @steve-chavez 

For Same db and same schema, tenants separated by RLS policies approach, does the auth table have unique email requirement ? 
I am building a platform where it's very possible that a user can belong to multiple tenants.

**Reply by w3b6x9** [reply]
@KhaledGabr 

> does the auth table have unique email requirement ?

yes, there's a unique index on the email column.

> I am building a platform where it's very possible that a user can belong to multiple tenants.

I recommend you have another table that references the `auth.users`  table in this case.

**Reply by elhe26** [reply]
@w3b6x9 But how are we going to perform a comparison between Auth Table and the "Tenant User" Table? The issue here is that the user would be created one time for 1 tenant but what happens when the client gets created for a second tenant? Email is the same but the password may be different.

**Reply by kangmingtay** [reply]
hey @elhe26, yeah you're right, we currently don't support multi-tenancy on the auth schema due to the `UNIQUE` constraint on the `email` column.

**Reply by yzx396** [reply]
@kangmingtay is there any plan to add support of mult-tenancy on the auth schema?

**Reply by ausmurp** [reply]
Any multi-tenant app I've used still has the unique email constraint. Think about Slack, you can sign in with your personal email, and see all of your workspaces. The architecture for this is simple, you just have the following schema:

`accounts`, `account_roles`, and a `account_users` (account_id, user_id, account_role_id).

Then you require account_id (tenant_id, or whatever you want to call it) as a header for every request.

**Reply by elhe26** [reply]
@ausmurp The issue here is that the password may be different and if we use RLS + Auth Table the user won't be able to "sign up" since he already exists on Auth. 

For example, take Firebase Auth. As of this moment, this is the best multi-tenancy implementation I've seen so far. Easy to implement and deploy, you don't have to worry about multiple tables. You use the `tenantId` and the client can be subscribed to that tenant without any issues. If the client goes to a different tenant, the same procedure is implemented. 

@kangmingtay Do you have any plans on how to tackle this implementation? If so, where can I follow this?

Edit: 

Just following up on @yangzhaox's question.

**Reply by baderdean** [reply]
Hello,

in multi-tenancy we could have multiple levels: organization and users. The same users could belong to many organization. So many possible logins, one with Social Login for his NGO, one with his email address for his Gym, another with email address (the same one) for his children's School.

regards

**Comment by Zyles** [comment]
We need multi tenancy with one DB per customer.

Are there plans to add this?

**Comment by kangmingtay** [comment]
Hi everyone, we do plan to make gotrue multi-tenant eventually but we don't have a clear timeline for this yet. The upstream version of gotrue from netlify supports multi-tenancy using method (3) described by steve above: Same db and same schema, tenants separated by `instance_id`. However, we would prefer to either go with (1) or (2) and this would require quite a significant architectural change and not something we would like to rush into. Ideally, we would also want to dogfood it on our platform first before releasing it to the rest of the projects on Supabase!

**Reply by KhaledGabr** [reply]
This is the best news I have heard for a while! 

I would vote for the second approach (Same DB, different Schema) for the following reasons. 

1) It would be possible to make queries across multiple schemas  ( Useful for super admins )
2) More Cost effective I would imagine compared to separate database. 
3) No database table changes required (ex: adding instance_id to all tables). Meaning that any application can be converted to multi-tenant without much hassle.

**Reply by ADO** [reply]
I suspect this second approach of Same DB, different Schema (with a shared structure within those schemas in most use-cases) would not only have substantial security and import/export benefits by effectively siloing company/tenant data by schema but could potentially be a game-changer for realtime performance and improve scaling.

- For multi-tenant scenarios (e.g. company-level tenant accounts with multiple users who may potentially belong to other company/tenant accounts), queries are often company-wide data shared between users, i.e. merely restricted to the tenant. Think "public within the scope of the company"
- Within a tenant's own schema it would mean no extra RLS policies on those tables would be needed
- And you still have RLS if needed for "private to the user even within the scope of the company" anyway

Having different schemas per tenant would allow more (automatically scoped by schema) table-wide non-RLS subscriptions for realtime (faster), so 

- `socket.channel('realtime:public:table_here:tenant_id=tenant_id_here')` becomes:
- `socket.channel('realtime:tenant_schema:table_here')`

Even the RLS subscriptions _and_ subscriptions to filtered data should be way faster with the schema scoping.

I believe this could also _dramatically_ reduce [the RLS performance hit seen as subscribers grow](https://supabase.com/blog/realtime-row-level-security-in-postgresql#performance)... as it would be restricted to the scope of an individual schema/tenant's subscribers so that changes to one tenant's data don't require looking up the effects on every subscriber system-wide.

**Reply by victorsoliveira** [reply]
@kangmingtay 

How can i write in `instances` table and reference `instance_id` at `users` table ? I didnt found how to in flutter client api for example.

![image](https://user-images.githubusercontent.com/1494145/216147216-68db976c-e9f8-4e74-8b7a-4cd819c3212e.png)

Thanks!

**Reply by kangmingtay** [reply]
you can't right now because multi-tenancy is not supported on supabase auth yet - we don't plan to use the `instances` table too and we are going to be deprecating it

**Reply by victorsoliveira** [reply]
Ok @kangmingtay! Thanks for the information.

**Reply by barbinbrad** [reply]
If we set a header manually in the supabase-js client (a tenant id), is there any way to access to from RLS?

**Comment by btelles** [comment]
Great to hear that this has been considered. My vote would be for Option 1 above for the following reasons:

- Tenants may be hesitant to upgrade to the latest backend services, and any breaking changes in APIs would be unwelcome.
- Billing on a per-tenant basis would be slightly easier.

**Comment by rubenbase** [comment]
Hey team! Any place where we can track the multi-tenancy implementation process?

**Comment by bwinkers** [comment]
Option 1 would allow me to provision my tenants into DBs and the client could provision their sub-tenants in schemas. All while keeping data in smaller packages that are easier to move, backup and restore.

**Comment by rchrdnsh** [comment]
Any updates that can be shared on this subject? Would love to know :-)

**Comment by ghost** [comment]
For Option 1, separate db per tenant - Could you just use the new [Management API](https://supabase.com/docs/reference/api/create-a-project) to create a new project for each tenant and then deploy to it?

I was thinking of trying something like that and then I would just have to figure out how to let users switch between organizations if they belong to multiple orgs. One "low rent" way I thought of for doing this: [generate a magic link](https://supabase.com/docs/reference/javascript/auth-admin-generatelink) for the org they want to switch to and navigate them to it.

I think this solution would require:

- A central project for orchestrating your SaaS concerns, tracking which tenant owns which Supabase project, etc.
- A separate domain/subdomain per tenant frontend OR carefully controlled creation of supabase-js clients.
- A way for each tenant frontend to safely ask for the list of other tenants a user belongs to and switch to them. (Could let each tenant frontend talk to the central SaaS project, but we don't want to do it so we can figure this out on the backend by letting tenant Edge functions use supabase-js to work with the central project with a service key.)

To be clear: Tenants would not own their Supabase projects in any way as developers or have any control over them such as rolling out their own Edge functions. They're just users.

**Comment by Sun3** [comment]
I am also interested in Multi-tenant...

**Comment by peacocc** [comment]
Hi. I am trying to build a multi tenancy platform and have run into the issues outlined above. After trying Next-Auth, which only has a single email call back location, I was hoping that supabase would at least allow me to programatically change the siteUrl property via the Management API so I can send the email confirmation to a custom tenant domain. Is there any news from the Supabase team on whether a solution will be provided on any of the options above?

**Comment by elhe26** [comment]
Is there any updates to this request? 

This request along with transaction support are the only things stopping our company from phasing out firebase completely.

**Comment by wixys** [comment]
Hi everyone, any updates about Multi-tenant feature? 👀

**Comment by liesahead** [comment]
Tough choice, but I made it in favor of firebase because of the multitenancy.

**Comment by mosnicholas** [comment]
+1 this would be helpful :)

**Comment by sebastianbachmann** [comment]
Not really sure, but is this new Supavisor thing going in the right direction? https://github.com/supabase/supavisor

**Comment by eibrahimov** [comment]
+1

**Comment by larksoftware** [comment]
+1

**Comment by wandersonalwes** [comment]
+1

**Comment by christopheragnus** [comment]
+1

**Comment by aaronguostudio** [comment]
+1

**Comment by DelightfulMoose6158** [comment]
+1

**Reply by rzzo** [reply]
+1

**Comment by Fibtegis** [comment]
+1

**Comment by mrpotato3** [comment]
+1

**Comment by Pranit-Harekar** [comment]
+1  I am interested in the separate schema same db approach. Is there any progress? Any visibility into the timeline would really help me in making decisions. Thanks!

**Reply by FlexiFarmAdmin** [reply]
Check out [Supabase Custom Claims](https://github.com/supabase-community/supabase-custom-claims/tree/8f9aca95845f8861bd6f0be086cc36c15ea0ff99) - we are using it for the one-schema-per-account approach with users having multiple account memberships (akin to Slack). 

Added benefit: Supabase realtime listeners become __far more performant__ for several reasons:

- affected changed rows are automatically scoped within the schema 
- client listeners should only be subscribing to table(s) within the current UI account schema
- in our case there is no further per-user scoping but if there was, it would be RLS within the account which is a whole lot better than system-wide!

__Beware that the Supabase UI itself is currently not able to handle a significant number of schemas with many tables in each__ (I am unsure of the threshold but it was certainly under 1000 account schemas) as it preload ALL schemas with ALL tables whenever you visit several areas of the UI (eg table editor, etc) and __will fail/timeout every time.__ I raised this with support but they have no plans to fix this at least anytime soon and pointed me towards doing more via PGAdmin or equivalent. For example whenever a new account is created and the schema is created as part of a callback server-side db migration, we 

- create the schema
- add the table structure (and default seeded content) within the schema
- automatically apply Row Level Security
- grant schema `usage` and `select` to `authenticated`
- add a policy to enable realtime on relevant tables

And whenever a user has a membership to an account, `SELECT set_claim('user-id-here', 'schema-here', 'true')`

**Reply by anilbeyazoglu** [reply]
Hi @FlexiFarmAdmin, we are using the same approach but to be able to use use supabase client with different schema we have to add it to Exposed Schemas in API settings manually. How do you handle that?

**Reply by FlexiFarmAdmin** [reply]
Hi @anilbeyazoglu - have a look at [Using Custom Schemas](https://supabase.com/docs/guides/api/using-custom-schemas) - we run a subset of that SQL direct to the Postgres instance (not via the API or UI) automatically when a new account is created, see the last three bullet points in my previous post. That should open up the schema for use via the API and Realtime.

**Reply by n84ck** [reply]
Hi @FlexiFarmAdmin, If I understand correctly, the same tables exist in different schemas for each tenant. How did you solve the data querying in the client application (e.g NEXT.js)? Is there a way to handle this in a typed manner, possibly with some kind of ORM?

**Reply by FlexiFarmAdmin** [reply]
Hi @n84ck - that's correct; one schema per client, with the same table structure in each schema. Our `public` schema has all those tables too, they're just empty (and some extra genuinely shared populated tables in `public` too). Which means types can be inferred, for example with a `prisma db pull` or equivalent.

I have found some ORMs quite inflexible in terms of specifying a schema which is often expected to be public and only public, or at the very least not dynamic. We're using SvelteKit and we use our own API to connect to the supabase connection pooler direct nowadays but originally and conceptually it was working using the supabase client and just specifying the schema name there explicitly.

**Comment by carloslato** [comment]
+1

**Comment by onursagir** [comment]
It's sub ideal but I've managed to get this working by saving tenants with their name and a uuid in the database and then prefixing all emails with the said uuid. However I've only tested this for email auth idk if this would for SSO. I am sure that something similar can be done to make phone authentication work tho. I've pushed my proof of concept repo to Github and it can be seen [here](https://github.com/onursagir/supabase-multi-tenancy)

**Comment by MaximusMcCann** [comment]
+1

**Comment by akshykhade** [comment]
+1

**Comment by Sun3** [comment]
Any updates on the matter, for enterprise level applications this feature is a must.

Thanks.

**Reply by RepairYourTech** [reply]
Why are you using a community edition for enterprise level applications lol. Grift much?

**Comment by Sun3** [comment]
ps: I am trying to convert my apps from Firebase to Supabase and create new apps with Supabase. Firebase auth as tenantId built-in. I believe this would be beneficiary to Supabase to support these kind of scenarios.

For my apps each tenant is a separate company and users do not share their email between tenant. At the worst for scenarios that would need this feature, they can use different emails for each tenant.

Thanks.

**Reply by wixys** [reply]
In searching for solutions for multi-tenant systems, I came across this: https://www.thenile.dev/

**Reply by Sun3** [reply]
I'll check it out...

**Reply by Sun3** [reply]
Looks very interesting...

**Comment by Pramuspl** [comment]
+1. Multi-tenant auth is a must-have. I've been looking for info a couple of days until I reached this thread.

I'd be great if you had some kind of a general roadmap, like Remix project does for example.

**Comment by KieranFleckney** [comment]
1+

**Comment by viktorsmakreckis** [comment]
plus one

**Comment by rsbichkar** [comment]
+1

**Comment by siddiqss** [comment]
+1

**Comment by Bych** [comment]
+1

**Comment by reedwane** [comment]
+1

**Comment by Mihai-github** [comment]
+1

**Comment by josecaputo** [comment]
Hello @kangmingtay 

Do you have any update or eta about this feature to share with us? 

It would be awesome to have this as a built-in solution as I have no code experience and it seems a bit complicated and confusing to me to implement what was suggested for schemas. 

Appreciate your attention

**Comment by josecaputo** [comment]
Hi @FlexiFarmAdmin,

Your solution seems to be what I am looking for however I have no code experience and I was wondering if you could send me a quote to prepare a step-by-step guide so I can implement same solution in supabase/schemas to manage my companies data's.

Thank you very much

**Comment by wandersonalwes** [comment]
+1

**Comment by minimalknave127** [comment]
+1

**Reply by coheedev** [reply]
+1

**Comment by slavakurilyak** [comment]
Check out [supa.guide](https://supa.guide/) by @activenode which shows you how to build a multi-tenant platform

**Reply by Zyles** [reply]
> Check out [supa.guide](https://supa.guide/) by @activenode which shows you how to build a multi-tenant platform

If you are going to promote your book you should write on the page what type of technique the multi tenancy is using.

I wouldn't pay for any book that tells me to put tentant_id column on every row.

**Reply by activenode** [reply]
Hey, author here (thanks for tagging). 

> I wouldn't pay for any book that tells me to put tentant_id column on every row.
I'm happy to help you if the book is for you but I'm not sure what you would expect. There are pretty much 3 ways of doing multi-tenancy. 

1. Separate DB (which is basically separate Supabase project). No-brainer, can also be automated, although I wouldn't want to spin up a new DB for each client - but really depends on what you want.
2. Same DB but different schema -> "just" adapt the schema in the initiated Supabase client. There are a few other things to manage though given that the built-in mailing would need to be adapted which is why I don't recommend using the built-in mailing in my book when using multi-tenancy
3. Same schema but proper DB architecture, same as in No2 it applies you want to separate E-Mails etc. as well and this is what's done in the book and yes, this requires also (not just) adding logically the tenant id to the rows.

Scrolling up I now see that @steve-chavez already gave that similiar answer in the beginning so I mine are just adding a bit more context here.

So question remains: What do you expect to see? @Zyles 

I can't elaborate the whole book but let me tell you there's certainly so much more than just "adding a tenant id". E.g. optimizing the RLS accordingly, emptying the built-in templates so abusive use (forcing to use the built-in one instead of the custom one) is avoided, etc. . The book has 500+ pages so it's really hard to give a summary here

**Reply by barbinbrad** [reply]
I'd like to do 2, but I have to do 3, because even though the client supports schemas, I haven't figured out how to tackle migrations. Ideally I'd have one set of migrations that'd fan out to multiple schemas.

**Reply by Zyles** [reply]
Different schema per tenant is the minimum acceptable solution.

Also different database per tenant depending on what type of information you're storing (regulations).

**Reply by activenode** [reply]
> Also different database per tenant depending on what type of information you're storing (regulations).

That's the definition of "new Supabase project for each client" so that's already available and you can automate it with the API: https://supabase.com/docs/reference/api/v1-create-a-project 


> Different schema per tenant is the minimum acceptable solution.
Given that your database has different extensions who will also live in different schemas as well as given the auth schema is its own schema apart from "tenant" schemas, I'm not seeing how "the minimum acceptable solution" provides any security benefit as it really does not. 

I can pretty much pull up a visionary glass ball and state: You will never see this implemented in Supabase and I also don't know of any competitor who does that this way. I certainly will you not recommend my book since you are already primed into a direction that I find hard to find consensus with but there's still the little hope in me that reading it would maybe change your mind about the necessity of schema separation.

Tldr: Schema separation does not necessarily provide any added security benefit as both can implement role-based security by all means. So all that remains is my understanding that you want it to simply go to a schema and only see specific tenant data there without the need to filter, that I could kinda understand. But in this case you're best of using the "Create new project" API around your infrastructure.

**Reply by ammelto** [reply]
What about when different tenants have slightly different table structures? I'm working on an application that has a ton of customization for each tenant, some even include properties in data models which might be entirely unique to that tenant. In our case, security and separation are important but so is the ability to provide a high degree of flexibility on the underlying data structures on similarly names tables.

**Reply by FlexiFarmAdmin** [reply]
@ammelto different table structures would have to be separate database or at least schemas. For separate schemas you can achieve that [as above](https://github.com/orgs/supabase/discussions/1615#discussioncomment-6632780) - which is what we are using in our beta atm (even though each client currently has the same structure that won't necessarily continue like that forever so it's flexible)

**Reply by ammelto** [reply]
@FlexiFarmAdmin Oh wow this is fantastic! I must have missed your post when skimming through the comments here. This is exactly what I was looking for, thanks!

**Comment by rajivkr** [comment]
Is there any update on this? We have multiple users who use a variety of login methods across tenants, like google for tenant 1, github for tenant 2 etc. So, ideally we should allow users to signup with any means across tenants.

**Reply by activenode** [reply]
For real, if you not only share database structure but different configurations, there is no other answer than pulling up a Supabase project for each client and building your automation around it with stuff like proper seeds and migration files (hint: https://www.youtube.com/watch?v=nyX_EygplXQ) .

Project Platform API: https://supabase.com/docs/reference/api/v1-create-a-project


If it's "only" about the different Login, that can be easily solved with an added identity_to_tenant table and then forwarding properly.

**Comment by RickVM** [comment]
Would be awesome to have the option to dynamically set the 'from' email when sending authentication emails. We have a multi-tenant setup going with supabase auth, but are restricted to sending our sign in emails from 1 address though we'd like to change the domain based on the organization/domain that is active.

**Reply by activenode** [reply]
Use the Auth Email Hook, that's the best practice! https://supabase.com/docs/guides/auth/auth-hooks/send-email-hook?queryGroups=language&language=sql 

Full flexibility, usable with Edge Functions as well.

Cheers, [activeno.de ](https://activeno.de)

**Reply by RickVM** [reply]
Awesome, seems like exactly what I need. Thanks! @activenode

**Comment by rh-id** [comment]
I think it is possible currently. Using http server like nginx to route request based on apikey header. So you have a list of apikey mapped to supabase instances.

It is not one schema one tenant but one supabase instance one tenant

**Comment by benkissi** [comment]
Can something be learnt from this https://neon.com/docs/guides/multitenancy ?

**Reply by RepairYourTech** [reply]
The problem is supabase and postgres itself is not meant to be miltitenant.  

Thankfully lots of nice new databases are evolving now, like SurrealDB
